Checks/PWR079/README.md

```diff
@@ -0,0 +1,336 @@
+
+Some compilers may appear to "help" by zero-initializing variables under
+certain conditions (e.g., specific compilation flags). While this can make
+technically incorrect code run as originally intended, relying on such
+incidental behavior creates a false sense of security and masks underlying
+logical errors. Ultimately, this can vanish under different compilation
+settings and can vary between compilers.
+
+Lastly, while some compilers provide options to automatically initialize
+certain data types (e.g., `gfortran`'s `-finit-integer=<value>` or `gcc`'s
+`-ftrivial-auto-var-init=<value>`), these features reduce portability to other
+development environments and hide problems in the code rather than addressing
+them.
+
+### Code examples
+
+#### C
+
+Consider the following code, which aims to sum the elements of an array:
+
+```c
+// example_array.c
+#include <stdio.h>
+
+__attribute__((pure)) double sum_array(double *array, int size) {
+  double sum;
+
+  for (int i = 0; i < size; ++i) {
+    sum += array[i];
+  }
+
+  return sum;
+}
+
+int main() {
+  double array[] = {0.24, 0.33, 0.17, 0.89, 0.05};
+  double sum = sum_array(array, 5);
+  printf("Sum is: %f\n", sum);
+
+  return 0;
+}
+```
+
+Note how `sum`, an automatic variable, is never explicitly initialized.
+Although it might seem like it "should" logically start at `0`, the C standard
+does not guarantee this. Thus, the initial value of `sum` is indeterminate,
+leading to different outcomes depending on the compiler and its settings:
+
+- For instance, `gcc -O2` appears to start `sum` at `0`, allowing the program
+  to work as intended:
+
+```txt
+$ gcc --version
+gcc (Debian 14.2.0-8) 14.2.0
+$ gcc -O2 example_array.c -o example_array_gcc
+$ ./example_array_gcc 
+Sum is: 1.680000
+```
+
+- However, with `clang -O2`, `sum` appears to contain arbitrary data, leading
+  to incorrect results:
+
+```txt
+$ clang --version
+Debian clang version 19.1.5 (1)
+$ clang -O2 example_array.c -o example_array_clang
+$ ./example_array_clang 
+Sum is: nan
+```
+
+The solution is straightforward, always initialize variables before using them:
+
+```c
+double sum = 0.0;
+```
+
+This principle applies to all variable types, including other elemental types
+like `int`, `struct` elements, and pointers.
+
+Pointers are particularly important in C, as they are commonly used to
+represent n-dimensional arrays. However, uninitialized pointers can easily lead
+to invalid memory accesses and program crashes. Let's consider another example
+code with pointers, where a computational function ensures the received
+pointers are valid before accessing their contents:
+
+```c
+// example_matrix.c
+#include <stdio.h>
+
+void perform_computation(double *matrix_a, double *matrix_b) {
+  if (matrix_a == NULL || matrix_b == NULL) {
+    printf("A matrix is NULL; skipping computation\n");
+    return;
+  }
+
+  printf("Performing computation...\n");
+}
+
+int main() {
+  double *matrix_a, *matrix_b;
```

Comment:
Consider highlighting this line in the code for clarity! You can follow a similar approach to the example provided in https://open-catalog.codee.com/Checks/PWR068/#code-examples.

Reply:
Done!